### DISCUSSION TITLE
Supporting timestamp in prometheus exporter

### DISCUSSION THREAD

**Discussion by kotharironak** [open-telemetry/opentelemetry-java#3696] [Q&A]
As per otel metrics proto, each metric's NumberDataPoint has a timestamp - `[time_unix_nano](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L376)` - for recording the moment when the observation was aggregated.

But, in the Prometheus-exporter, this value is ignored while creating samples (link of [createSample](https://github.com/open-telemetry/opentelemetry-java/blob/8d7cca7caea1a322d6cee8abef6fa49fe9a3c0e4/exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/MetricAdapter.java#L272) function). In my understanding, it is designed from sdk point of view and use in the application, so it will be a scrap moment when these metrics were collected.

I am looking into Kafka sink kind of a job where otlp metrics converted Prometheus format. I would like to re-use this exporter library. In that context, is there a way to support the `time_unix_nano` in this library while exporting? Can we make this library generic?

e.g Should we extend the signature of createSample to take timestamp?
```
Sample createSample(
    String name,
    List<String> labelNames,
    List<String> labelValues,
    double value,
    @Nullable ExemplarData exemplar,
long epochNanos)
```

**Comment by kotharironak** [comment]
@jsuereth what do you think?

**Reply by jsuereth** [reply]
As you've seen in that code, the exporter didn't intend to expose itself as anything but a normal prometheus exporter.   As you saw, we assume that the prometheus collection timestamp *is* the same as the `time_unix_nano` we would have reported anyway.


How do you want to re-use this library?   Are you calling the public methods on the prometheus exporter?   Are you calling individual conversion methods?   What is it you'd like exposed?

**Reply by kotharironak** [reply]
>How do you want to re-use this library?

I am intent to use by creating an instance of `PrometheusCollector` by providing `MetricProducer`.  I have a pipeline that receives metrics in otlp format,  which are enriched with a few new labels in the process. Once, the enrichment is done, I need to export them to Prometheus. Here, I would like to reuse this library. So, I have written a MetricProducer. But, those metrics already have a timestamp (generated at the client end).
> Are you calling the public methods on the prometheus exporter?

no

> Are you calling individual conversion methods?

no

> What is it you'd like exposed?

I am expecting that it should support timeStamp from MetricData points (PointData -> `long getEpochNanos();`).
Something like this - https://github.com/open-telemetry/opentelemetry-java/pull/3700
